### PR TITLE
update checkout from disk instead of memory buffer.

### PR DESCRIPTION
**Pull request by SidW**
Added second checksum validation after hdfsFetcher had pulled data to local disk. The fetcher now exams checksum values twice:
1. when the data reaches memory buffer.
2. when the data is written in hard disk. 

I didn't remove the memory buffer check since it kept writing checksum data into HdfsCopyStats while data transition.

### CODE REVIEW

File: contrib/hadoop-store-builder/src/java/voldemort/store/readonly/fetcher/BasicFetchStrategy.java [hunk 1]
```diff
@@ -208,6 +216,26 @@ public BasicFetchStrategy(HdfsFetcher fetcher,
                 }
             }
         }
+
+        //second time checksum validation. Check if the local file is consistent with the buffer
+        if (bufferCheckSumGenerator != null){
+            CheckSum fileCheckSum = CheckSum.getInstance(checkSumType);
```
Comment (arunthirupathi) [inline contrib/hadoop-store-builder/src/java/voldemort/store/readonly/fetcher/BasicFetchStrategy.java]: This should probably be called fileCheckSumGenerator.

File: contrib/hadoop-store-builder/src/java/voldemort/store/readonly/fetcher/BasicFetchStrategy.java [hunk 2]
```diff
@@ -208,6 +216,26 @@ public BasicFetchStrategy(HdfsFetcher fetcher,
...
+            BufferedInputStream in = new BufferedInputStream(new FileInputStream(dest));
+            int read;
+
+            try {
+                while ((read = in.read(buffer)) >= 0) {
+                    fileCheckSum.update(buffer, 0, read);
+                }
+
+                if (ByteUtils.compare(checkSum, fileCheckSum.getCheckSum()) != 0)
+                    throw new VoldemortException("Local file: " + dest.getAbsolutePath() +
```
Comment (arunthirupathi) [inline contrib/hadoop-store-builder/src/java/voldemort/store/readonly/fetcher/BasicFetchStrategy.java]: Can you also print what is the fileCheckSum and buffer checksum in the error logs ?

File: contrib/hadoop-store-builder/src/java/voldemort/store/readonly/fetcher/BasicFetchStrategy.java [hunk 3]
```diff
@@ -1,11 +1,15 @@
...
 
+import java.io.BufferedInputStream;
 import java.io.BufferedOutputStream;
 import java.io.File;
+import java.io.FileInputStream;
 import java.io.FileOutputStream;
 import java.io.FilterInputStream;
 import java.io.IOException;
 import java.io.OutputStream;
+import java.nio.file.Files;
```
Comment (arunthirupathi) [inline contrib/hadoop-store-builder/src/java/voldemort/store/readonly/fetcher/BasicFetchStrategy.java]: Files and Paths can be removed, if they are not used.

**Comment by arunthirupathi** [comment]
looks good, will merge this in.

**Comment by arunthirupathi** [comment]
Rebased and committed. https://github.com/voldemort/voldemort/commit/64726a93da7e8f8a0175f5d073e751caf84c0c33

**Comment by FelixGV** [comment]
@SidW, @arunthirupathi,

I assume this change is to catch issues where the filesystem may fail silently by corrupting the data without reporting an error at write-time?

I have two related questions:

1. Did we measure the impact in terms of total fetch time?
2. What about IO, are we trashing the page cache? Conversely, if we suspect that there is silent data corruption going on in an SSD, is there a possibility that the OS will have a correct (non-corrupt) data file content in its page cache, and thus also silently fail the second check, since what Java thinks it is loading content from the SSD but it may actually be coming from the OS' page cache?

@gnb, do you have any thoughts on the above?

-F

**Comment by SidW** [comment]
1. I haven't measure the impact regarding to the time. 
2. Didn't even think about this issue and would know the truth :)

**Pull request closed without merging**